### DISCUSSION TITLE
Is there any way to obtain the full callstack under -O2/-O3/inline case?

### DISCUSSION THREAD

**Discussion by xiejingwei** [libunwind/libunwind#817] [Q&A]
Since tail call optimization will occur under -O2/-O3 compiling option，this seems to result in an incomplete callstack.

Inline functions seem to cause similar problems.

However, gdb backtrace seems to be able to produce complete callstack, including frames concerning inlined functions.

Is there a way to solve this?

**Comment by milianw** [comment]
you can use elfutils or similar to get all inline frames for a stack frame obtained by libunwind. I do this in heaptrack e.g. I also gave a talk on this topic at CppCon a few years ago, you can find more information here: https://github.com/milianw/how-to-write-a-memory-profiler

Hope that helps you

**Comment by bregma** [comment]
The difference between libunwind and gdb in performing backtraces is that libuwind uses the current IP to get the current frame info and that's it, for speed. GDB has the luxury of doing a reverse lookup of all function name strings mapped to an IP address range, which can be an order of magnitude slower but because of the luxury of doing offline processing, it doesn't matter so much.

It's a speed-vs-quantity tradeoff.
